tests/unit/test_partition_by_constraint.cpp

```diff
@@ -0,0 +1,187 @@
+            EXPECT_EQ(expected_indices[i], output_constraint.contiguous[i]);
+    }
+}
+
+TEST(partition_by_constraint, partition_independent) {
+    iarray input_index(input_size_);
+    iarray expected_indices;
+    multicore::constraint_partitions output_constraint;
+
+    for (unsigned i = 0; i < input_size_; i++) {
+        input_index[i] = i * 2;
+        if(i % simd_width_ == 0)
+            expected_indices.push_back(i);
+    }
+
+    multicore::generate_index_constraint_partitions(input_index, output_constraint, input_size_);
+
+    EXPECT_EQ(0, output_constraint.constant.size());
+    EXPECT_EQ(0, output_constraint.none.size());
+    if(simd_width_ != 1) {
+        EXPECT_EQ(input_size_/simd_width_, output_constraint.independent.size());
+        EXPECT_EQ(0, output_constraint.contiguous.size());
+
+        for (unsigned i = 0; i < input_size_ / simd_width_; i++)
+            EXPECT_EQ(expected_indices[i], output_constraint.independent[i]);
+    }
+    else {
+        EXPECT_EQ(0, output_constraint.independent.size());
+        EXPECT_EQ(input_size_/simd_width_, output_constraint.contiguous.size());
+
+        for (unsigned i = 0; i < input_size_ / simd_width_; i++)
+            EXPECT_EQ(expected_indices[i], output_constraint.contiguous[i]);
+    }
+}
+
+TEST(partition_by_constraint, partition_serial) {
+    iarray input_index(input_size_);
+    iarray expected_indices;
+    multicore::constraint_partitions output_constraint;
+
+    for (unsigned i = 0; i < input_size_; i++) {
+        input_index[i] = i / ((simd_width_ + 1)/ 2);
+        if(i % simd_width_ == 0)
+            expected_indices.push_back(i);
+    }
+
+    multicore::generate_index_constraint_partitions(input_index, output_constraint, input_size_);
+
+    EXPECT_EQ(0, output_constraint.independent.size());
+    EXPECT_EQ(0, output_constraint.constant.size());
+    if(simd_width_ != 1) {
+        EXPECT_EQ(input_size_/simd_width_, output_constraint.none.size());
+        EXPECT_EQ(0, output_constraint.contiguous.size());
+
+        for (unsigned i = 0; i < input_size_ / simd_width_; i++)
+            EXPECT_EQ(expected_indices[i], output_constraint.none[i]);
+    }
+    else {
+        EXPECT_EQ(0, output_constraint.none.size());
+        EXPECT_EQ(input_size_/simd_width_, output_constraint.contiguous.size());
+
+        for (unsigned i = 0; i < input_size_ / simd_width_; i++)
+            EXPECT_EQ(expected_indices[i], output_constraint.contiguous[i]);
+    }
+}
+
+TEST(partition_by_constraint, partition_random) {
+    iarray input_index(input_size_);
+    iarray expected_indices_contiguous, expected_indices_constant,
+            expected_indices_independent, expected_indices_serial;
+    multicore::constraint_partitions output_constraint;
+
+
+    const int c = 5;
+//shuffle here
+    for (unsigned i = 0; i < input_size_; i++) {
+        input_index[i] = (i < input_size_ / 4 ? i / ((simd_width_ + 1)/ 2) :
```

Comment:
This ternary expression can be made clearer by removing some parenthesis `()` and realignment, e.g.:
```
        input_index[i] = i<input_size_/4   ? i/((simd_width_ + 1)/ 2): 
                         i<input_size_/2   ? i*2:
                         i<input_size_*3/4 ? c:
                         i;
```